### DISCUSSION TITLE
TLS1.3: Understanding HKDF calls

### DISCUSSION THREAD

**Discussion by kshitizvars** [openssl-projects/pkcs11-provider#442] [Q&A] [answered]
Hi @simo5 @Jakuje 

In p11prov_exch_hkdf_derive API, we are calling EVP_KDF_derive (openssl API) to perform HKDF derive, shouldn't  it be like in below screenshot:-
![image](https://github.com/user-attachments/assets/94898a04-ef5e-48b1-8309-c242656fb6e8)

Correct if I am wrong here. 

@sahilnxp

**Comment by simo5** [comment]
We need to rework it to look like p11prov_ecdh_derive() more or less

**Reply by kshitizvars** [reply]
So, control will reach to token after calling EVP_KDF_derive?

**Reply by simo5** [reply]
I am not sure the current code works as it should

**Reply by simo5** [reply]
but eventually we should be calling pkcs11_DeriveKey()

**Comment by simo5** [comment, marked as answer]
Ok sorry looking at this more carefully I now remember how odd the code is around key derivation in openssl.

The key exchange providers "wrap" kdf providers, so in thie case calling EVP_KDF_derive(hkdfctx->kdfctx, secret, outlen, NULL) should be correct as this will do nothing more than simply end up calling the *KDF* provider with a KDF context for HKDF.

**Reply by simo5** [reply]
Also the context is guaranteed to be the PKCS#11 provider KDF context, see how kdfctx->kdfctx
is initialized in p11prov_exch_hkdf_newctx()

**Reply by kshitizvars** [reply]
So when EVP_KDF_derive() is called, we first go to OpenSSL and then from there we again come into provider which calls the C_DeriveKey(), Is my understanding correct?
I am not able to understand from where Token C_DeriveKey() will be called in this case?

**Reply by simo5** [reply]
Yes EVP_KDF_derive is given a context explicitly taken with provider=pkcs11 as property, so the only provider that will be selected in pkcs11.
EVP_KDF_derive() -> p11prov_hkdf_derive() -> p11prov_derive_key() -> p11prov_DeriveKey()

**Reply by kshitizvars** [reply]
I understand the flow now, since we are in the provider only, why we need to call **EVP_KDF_derive()**, can't we just directly call  **p11prov_hkdf_derive()** without calling EVP_KDF_derive()?

**Reply by simo5** [reply]
To avoid special casing, there is no real gain in multiplying paths

**Comment by kshitizvars** [comment]
Hi @simo5 @Jakuje 

We have added **CKM_HKDF_DERIVE** mechanism in our token, but we don't see **p11prov_exch_hkdf_derive** getting called when working with TLS1.3. Is there any config or any other option that we need to enable for offloading HKDF operation to provider.

**Reply by simo5** [reply]
As long as the mechanism is exposed in the mech lists pkcs11-provider will pick it up.
Of course, as usual you will have to configure OpenSSL to defer all operations to the token or ensure at least one of the keys is a pkcs11 key that the token will not allow to export, otherwise as usual OpenSSL will try to import the key in the default provider and use the default provider implementation in preference.

**Reply by kshitizvars** [reply]
As per my current understanding, for TLS1.3, OpenSSL calls **EVP_KDF_fetch** with **TLS13-KDF** as algo name, which seems missing in provider's code, hence it falls for OpenSSL default implementation whereas for running HKDF standalone like in below command, it calls **EVP_KDF_fetch** with algo name as **HKDF** :-

```
export HKDF_HEX_SECRET=ffeeddccbbaa
export HKDF_HEX_SALT=ffeeddccbbaa
export HKDF_HEX_INFO=ffeeddccbbaa
 
openssl pkeyutl -derive -kdf HKDF -kdflen 48 -pkeyopt md:SHA256 -pkeyopt mode:EXTRACT_AND_EXPAND -pkeyopt hexkey:${HKDF_HEX_SECRET} -pkeyopt hexsalt:${HKDF_HEX_SALT} -pkeyopt hexinfo:${HKDF_HEX_INFO} -out ${TMPPDIR}/hkdf1-out-pkcs11.bin -propquery provider=pkcs11'
```

Correct me if I am wrong here.

Btw, we are able to see CKM_HKDF_Derive mechanism in list-mechanism command, but still **p11prov_exch_hkdf_derive** is not getting called.

**Reply by simo5** [reply]
@kshitizvars Looks like we'll need to implement TLS13-KDF, can you open an issue about that ?

**Reply by simo5** [reply]
I think in the command you need to add '-provider pkcs11' before -propquery

**Reply by kshitizvars** [reply]
I have added below config related change in openssl.conf for forcing it on pkcs11-provider:-
```
[openssl_init]
providers = provider_sect
alg_section = algorithm_sect
 
# List of providers to load
[provider_sect]
default = default_sect
pkcs11 = pkcs11_sect
 
[default_sect]
activate = 1
[pkcs11_sect]
module = /usr/lib/ossl-modules/pkcs11.so
pkcs11-module-path = /usr/lib/libckteec.so.0
pkcs11-module-cache-keys = false
pkcs11-module-quirks = no-operation-state
pkcs11-module-block-operations = digest
activate = 1
 
[algorithm_sect]
default_properties = ?provider=pkcs11
```

**Reply by simo5** [reply]
and HKDF is still not use ?

**Reply by simo5** [reply]
note that there is a test in the testsuite that explicitly tests HKDF and I see in the generated p11prov-debug.log that the token's HKDF is used, so I am not sure why it does not work in your case.

See tests/thkdf

**Reply by kshitizvars** [reply]
Yes, I have taken the reference from **tests/thkdf**, these tests are for standalone HKDF testing, it is working fine, I am also able to see logs in debug log, but while we run TLS1.3 connection, we don't see any HKDF functionality getting called, and it is probably because of missing **TLS13-KDF**.

**Reply by simo5** [reply]
indeed if the name is not present the pkcs11 provider will never be used

**Reply by simo5** [reply]
Filed #445 to follow up on this
